src/index.ts

```diff
@@ -187,41 +189,140 @@ export class Logger {
 	}
 
 	/**
-	 * Turns an object array into a JSON string
+	 * Turns an object into a string 
+	 * @param item The item to be stringified
+	 * @param actualConfig The config to use for stringification
+	 * @param referenceArr An array of references to check for circular references
+	 * @param depth The depth of the current item (used for indentation and reference checking)
+	 * @returns A JSON-adjacent string of item
+	 **/
+	private stringifyItem(item: any, actualConfig: LoggerConfig, referenceArr: Array<[any, number]>, depth: number = 0): string {
+		const whitespace = actualConfig.tabs ? "\t" : "  ";
+		const lineStartWhitespace = whitespace.repeat(depth);
+		const newline = actualConfig.carriageReturn ? "\r\n" : "\n";
+		const itemSeperator = (actualConfig.multilineObjects ? (newline + lineStartWhitespace) : whitespace);
+		let result = "";
+		switch (typeof (item)) {
+			case "undefined":
+				result = "undefined";
+				break;
+			case "boolean":
+				result = item.toString();
+				break;
+			case "number":
+				result = (item as number).toFixed(item % 1 > 0 ? actualConfig.decimalDigits : 0);
+				break;
+			case "bigint":
+				result = item.toString();
+				break;
+			case "string":
+				// if the string is from the recursive call, 
+				// it means it's a stringified object, so we need to wrap it in quotes
+				result = depth!==0 ? `"${item}"` : item;
+				break;
+			case "symbol":
+				// TODO: Figure out how to stringify symbols, whatever that means
+				result = "[symbol]";
+				break;
+			case "function":
+				// differentiate function from class
+				{
+					let hasNonTrivialReferences = false;
+					for(let i=0;i<referenceArr.length;i++) {
+						if(referenceArr[i][0] === item && depth > referenceArr[i][1]) {
+							result = "[unserializable object]";
+							hasNonTrivialReferences = true;
+							break;
+						}
+					}
+					if(hasNonTrivialReferences) break;
+				}
+				referenceArr.push([item, depth]);
+				if(item.name[0].toUpperCase()===item.name[0]) {
+					result = `[Class ${item.prototype.constructor.name}]`;
+				} else {
+					result = `[Function ${item.name}]`;
+				}
+				break;
+			case "object":
+				if (item === null) {
+					// null is an object, so we need to check for it first
+					result = "null";
+					break;
+				}
+				// does this object have non-trivial references?
+				{
+					let hasNonTrivialReferences = false;
+					for(let i=0;i<referenceArr.length;i++) {
+						if(referenceArr[i][0] === item && depth > referenceArr[i][1]) {
+							result = "[unserializable object]";
+							hasNonTrivialReferences = true;
```

Comment:
I believe non-trivial would be better. This algorithm does not actually properly detect if there are cyclical references, rather, just quits out if it detects that the same object was already referenced a layer higher. I believe this is ok for our purposes, but an optional "thorough serialization" flag could be added later that does element sorting by dependency to log properly at the cost of speed.

Reply:
I will rework the algorithm to look for circular references up the current dependency chain, rather than global reference depth counting. This way it can double data in some places but that way they will definitely be circular